core/src/main/scala/org/apache/spark/Dependency.scala

```diff
@@ -144,12 +144,16 @@ class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
     _shuffleMergedFinalized = true
   }
 
+  def shuffleMergeFinalized: Boolean = {
```

Comment:
Yes I agree. With Adaptively fetching shuffle mergers, we need to check whether the `shuffleId` is `shuffleMergeFinalized` or not but that internally checks `shuffleMergeEnabled` which would be false since we didn't have enough mergers during the start of the stage. Which is why we need to have an additional method distinguishing whether shuffle merge is finalized vs whether shuffleMergeEnabled and then shuffleMergeFinalized whenever we try to figure which stage to be computed next. We need to prevent the next stage from starting if the previous stage shuffle merge finalization is yet to be completed ([here](https://github.com/apache/spark/blob/f6128a6f4215dc45a19209d799dd9bf98fab6d8a/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L731)).

Reply:
Let us split the state out into two variables - since they are fundamentally two things represented earlier by one boolean.

* `shuffleMergeAllowed`
* `shufflerMergeEnabled`

If `shuffleMergeAllowed` is `false`, then all push based shuffle paths are ignored - and is statically set at driver.
It defaults to `canShuffleMergeBeEnabled` - but can be explicitly disabled : for example, when missing tasks are being computed for a determinate shuffle after it has been finalized.
This would be set at driver before the stage starts - and can be dependent upon at executors to enable push based shuffle paths.

`shufflerMergeEnabled` can be set to `true` only if `shuffleMergeAllowed` is `true` and additional conditions are met - for example, sufficient mergers are available.
Earlier, this was determined statically when stage started - with this PR, this can become `true` adaptively after stage starts.
`shuffleMergedFinalized` can remain as it was defined earlier.

Thoughts ?

+CC @Ngone51, @otterc

Reply:
I refactored this in to `shuffleMergeAllowed` and `shuffleMergeEnabled` to have a clear distinction  where `shuffleMergeAllowed` controls the static level of knobs like the following:
1. `Is RDD barrier?`
2. `Can Push shuffle be enabled?`
3. `Disabling push shuffle for retry once the determinate stage attempt is finalized` etc.
and `shuffleMergeEnabled` will be checked only when `shuffleMergeAllowed` is true along with that if sufficient mergers are available then it becomes `true`.

Given all the above, we still require separate methods one for `isShuffleMergeFinalized` only checking for the `shuffleMergedFinalized` value and `numPartitions > 0`.  In the existing code, before the `ShuffleMapStage` starts we are checking `if (!shuffleMergeFinalized)` then only we are calling `prepareShuffleServicesForShuffleMapStage` but it is possible the previous stage never had `shuffleMergeEnabled` due to not enough mergers therefore even the retry also not be shuffle merge enabled, ideally this can be shuffle merge enabled if enough mergers are available. But for proceeding with the next stage, we need to check for `isShuffleMergeFinalizedIfEnabled` which is checking `if (shuffleMergeEnabled) isShuffleMergeFinalized else true`

Reply:
Yes, that is correct.
As you mentioned above,

- retry for an INDETERMINATE shuffle - `newShuffleMergeState` will set `_shuffleMergedFinalized` to false.
- retry for a DETERMINATE shuffle 
  - `mergerLocs.nonEmpty` (merge enabled) -> `_shuffleMergeFinalized` would be `true` -> `shuffleMergeAllowed` -> false for retry/reuse cases.
  - `mergerLocs.isEmpty` (merge disabled) -> `_shuffleMergeFinalized` would be `false` -> `shuffleMergeAllowed` -> `true` for retry/reuse cases.